File: bindings/nim/canvas.nim
## Kryon Canvas - Love2D Style API
##
## Provides Love2D-inspired immediate mode drawing functions
## that map to the enhanced C canvas implementation

import core_kryon, core_kryon_canvas, runtime

export core_kryon

# ============================================================================
# Types and Enums (matching C definitions)
# ============================================================================

type
  DrawMode* = enum
    Fill = 0
    Line = 1

  LineStyle* = enum
    Solid = 0
    Rough = 1
    Smooth = 2

  LineJoin* = enum
    Miter = 0
    Bevel = 1
    Round = 2

  BlendMode* = enum
    Alpha = 0        # Standard alpha blending
    Additive = 1     # Additive blending
    Multiply = 2     # Multiplicative blending
    Subtract = 3     # Subtractive blending
    Screen = 4       # Screen blending
    Replace = 5      # Replace (no blending)

  Point* = tuple[x: float, y: float]

# ============================================================================
# Constants
# ============================================================================

const
  PI* = 3.141592653589793

  # Predefined colors (matching C definitions)
  COLOR_RED*     = 0xFF0000FF'u32
  COLOR_GREEN*   = 0x00FF00FF'u32
  COLOR_BLUE*    = 0x0000FFFF'u32
  COLOR_WHITE*   = 0xFFFFFFFF'u32
  COLOR_BLACK*   = 0x000000FF'u32
  COLOR_YELLOW*  = 0xFFFF00FF'u32
  COLOR_CYAN*    = 0x00FFFFFF'u32
  COLOR_MAGENTA* = 0xFF00FFFF'u32
  COLOR_GRAY*    = 0x808080FF'u32
  COLOR_ORANGE*  = 0xFFA500FF'u32
  COLOR_PURPLE*  = 0x800080FF'u32

# ============================================================================
# Canvas Lifecycle and Global State Management
# ============================================================================

proc initCanvas*(width, height: int) =
  ## Initialize the global canvas system
  kryonCanvasInit(uint16(width), uint16(height))

proc shutdownCanvas*() =
  ## Shutdown the global canvas system
  kryonCanvasShutdown()

proc resizeCanvas*(width, height: int) =
  ## Resize the canvas
  kryonCanvasResize(uint16(width), uint16(height))

proc clearCanvas*() =
  ## Clear the canvas with background color
  kryonCanvasClear()

proc clearCanvas*(color: uint32) =
  ## Clear the canvas with specific color
  kryonCanvasClearColor(color)

proc clearCanvas*(r, g, b, a: int) =
  ## Clear the canvas with specific RGBA color
  kryonCanvasClearColor(rgb(r, g, b, a))

# ============================================================================
# Drawing State Management (Love2D Style)
# ============================================================================

proc setColor*(color: uint32) =
  ## Set current drawing color
  kryonCanvasSetColor(color)

proc setColor*(r, g, b: int) =
  ## Set current drawing color (RGB)
  kryonCanvasSetColorRgb(uint8(r), uint8(g), uint8(b))

proc setColor*(r, g, b, a: int) =
  ## Set current drawing color (RGBA)
  kryonCanvasSetColorRgba(uint8(r), uint8(g), uint8(b), uint8(a))

proc getColor*(): uint32 =
  ## Get current drawing color
  kryonCanvasGetColor()

proc setBackgroundColor*(color: uint32) =
  ## Set background color
  kryonCanvasSetBackgroundColor(color)

proc setBackgroundColor*(r, g, b, a: int) =
  ## Set background color (RGBA)
  kryonCanvasSetBackgroundColorRgba(uint8(r), uint8(g), uint8(b), uint8(a))

proc setLineWidth*(width: float) =
  ## Set line width in pixels
  kryonCanvasSetLineWidth(floatToFp(width))

proc getLineWidth*(): float =
  ## Get current line width
  fpToFloat(kryonCanvasGetLineWidth())

proc setLineStyle*(style: LineStyle) =
  ## Set line style
  kryonCanvasSetLineStyle(cast[ kryon_line_style_t ](ord(style)))

proc getLineStyle*(): LineStyle =
  ## Get current line style
  LineStyle(ord(kryonCanvasGetLineStyle()))

proc setLineJoin*(join: LineJoin) =
  ## Set line join style
  kryonCanvasSetLineJoin(cast[ kryon_line_join_t ](ord(join)))

proc getLineJoin*(): LineJoin =
  ## Get current line join style
  LineJoin(ord(kryonCanvasGetLineJoin()))

proc setFont*(fontId: int) =
  ## Set current font by ID
  kryonCanvasSetFont(uint16(fontId))

proc getFont*(): int =
  ## Get current font ID
  int(kryonCanvasGetFont())

proc setBlendMode*(mode: BlendMode) =
  ## Set blend mode
  kryonCanvasSetBlendMode(cast[ kryon_blend_mode_t ](ord(mode)))

proc getBlendMode*(): BlendMode =
  ## Get current blend mode
  BlendMode(ord(kryonCanvasGetBlendMode()))

# ============================================================================
# Transform System (Love2D Style)
# ============================================================================

proc origin*() =
  ## Reset transform to identity
  kryonCanvasOrigin()

proc translate*(x, y: float) =
  ## Translate the coordinate system
  kryonCanvasTranslate(floatToFp(x), floatToFp(y))

proc rotate*(angle: float) =
  ## Rotate the coordinate system (angle in degrees)
  kryonCanvasRotate(floatToFp(angle))

proc rotateRad*(angle: float) =
  ## Rotate the coordinate system (angle in radians)
  kryonCanvasRotate(floatToFp(angle * 180.0 / PI))

proc scale*(sx, sy: float) =
  ## Scale the coordinate system
  kryonCanvasScale(floatToFp(sx), floatToFp(sy))

proc shear*(kx, ky: float) =
  ## Shear the coordinate system
  kryonCanvasShear(floatToFp(kx), floatToFp(ky))

proc push*() =
  ## Push current transform state onto the stack
  kryonCanvasPush()

proc pop*() =
  ## Pop transform state from the stack
  kryonCanvasPop()

# ============================================================================
# Basic Drawing Primitives (Love2D Style)
# ============================================================================

proc rectangle*(mode: DrawMode, x, y, width, height: float) =
  ## Draw a rectangle (filled or outline)
  kryonCanvasRectangle(
    cast[ kryon_draw_mode_t ](ord(mode)),
    floatToFp(x), floatToFp(y),
    floatToFp(width), floatToFp(height)
  )

proc rectangle*(mode: DrawMode, x, y, width, height: int) =
  ## Draw a rectangle (filled or outline) - integer version
  rectangle(mode, float(x), float(y), float(width), float(height))

proc circle*(mode: DrawMode, x, y, radius: float) =
  ## Draw a circle (filled or outline)
  kryonCanvasCircle(
    cast[ kryon_draw_mode_t ](ord(mode)),
    floatToFp(x), floatToFp(y),
    floatToFp(radius)
  )

proc circle*(mode: DrawMode, x, y, radius: int) =
  ## Draw a circle (filled or outline) - integer version
  circle(mode, float(x), float(y), float(radius))

proc ellipse*(mode: DrawMode, x, y, rx, ry: float) =
  ## Draw an ellipse (filled or outline)
  kryonCanvasEllipse(
    cast[ kryon_draw_mode_t ](ord(mode)),
    floatToFp(x), floatToFp(y),
    floatToFp(rx), floatToFp(ry)
  )

proc ellipse*(mode: DrawMode, x, y, rx, ry: int) =
  ## Draw an ellipse (filled or outline) - integer version
  ellipse(mode, float(x), float(y), float(rx), float(ry))

proc polygon*(mode: DrawMode, vertices: openArray[Point]) =
  ## Draw a polygon (filled or outline)
  if vertices.len == 0: return

  # Convert vertices to fixed-point array
  var fpVertices = newSeq[kryon_fp_t](vertices.len * 2)
  for i, vertex in vertices:
    fpVertices[i * 2] = floatToFp(vertex.x)
    fpVertices[i * 2 + 1] = floatToFp(vertex.y)

  kryonCanvasPolygon(
    cast[ kryon_draw_mode_t ](ord(mode)),
    fpVertices[0].addr, uint16(vertices.len)
  )

proc line*(x1, y1, x2, y2: float) =
  ## Draw a line between two points
  kryonCanvasLine(floatToFp(x1), floatToFp(y1), floatToFp(x2), floatToFp(y2))

proc line*(x1, y1, x2, y2: int) =
  ## Draw a line between two points - integer version
  line(float(x1), float(y1), float(x2), float(y2))

proc line*(points: openArray[Point]) =
  ## Draw connected lines through multiple points
  if points.len < 2: return

  # Convert vertices to fixed-point array
  var fpVertices = newSeq[kryon_fp_t](points.len * 2)
  for i, point in points:
    fpVertices[i * 2] = floatToFp(point.x)
    fpVertices[i * 2 + 1] = floatToFp(point.y)

  kryonCanvasLinePoints(fpVertices[0].addr, uint16(points.len))

proc point*(x, y: float) =
  ## Draw a single point
  kryonCanvasPoint(floatToFp(x), floatToFp(y))

proc point*(x, y: int) =
  ## Draw a single point - integer version
  point(float(x), float(y))

proc points*(points: openArray[Point]) =
  ## Draw multiple points
  if points.len == 0: return

  # Convert vertices to fixed-point array
  var fpVertices = newSeq[kryon_fp_t](points.len * 2)
  for i, point in points:
    fpVertices[i * 2] = floatToFp(point.x)
    fpVertices[i * 2 + 1] = floatToFp(point.y)

  kryonCanvasPoints(fpVertices[0].addr, uint16(points.len))

proc arc*(mode: DrawMode, cx, cy, radius: float, startAngle, endAngle: float) =
  ## Draw an arc (angle in radians)
  kryonCanvasArc(
    cast[ kryon_draw_mode_t ](ord(mode)),
    floatToFp(cx), floatToFp(cy),
    floatToFp(radius),
    floatToFp(startAngle), floatToFp(endAngle)
  )

proc arc*(mode: DrawMode, cx, cy, radius: int, startAngle, endAngle: float) =
  ## Draw an arc (angle in radians) - integer position/radius version
  arc(mode, float(cx), float(cy), float(radius), startAngle, endAngle)

# ============================================================================
# Text Rendering (Love2D Style)
# ============================================================================

proc print*(text: string, x, y: float) =
  ## Print text at the specified position
  kryonCanvasPrint(text.cstring, floatToFp(x), floatToFp(y))

proc print*(text: string, x, y: int) =
  ## Print text at the specified position - integer version
  print(text, float(x), float(y))

proc printf*(text: string, x, y: float, wrapLimit: float, align: int = 0) =
  ## Print formatted text with word wrapping and alignment
  kryonCanvasPrintf(text.cstring, floatToFp(x), floatToFp(y), floatToFp(wrapLimit), int32(align))

proc printf*(text: string, x, y: int, wrapLimit: int, align: int = 0) =
  ## Print formatted text with word wrapping and alignment - integer version
  printf(text, float(x), float(y), float(wrapLimit), align)

proc getTextWidth*(text: string): float =
  ## Get the width of text in pixels
  fpToFloat(kryonCanvasGetTextWidth(text.cstring))

proc getTextHeight*(): float =
  ## Get the height of current font in pixels
  fpToFloat(kryonCanvasGetTextHeight())

proc getFontSize*(): float =
  ## Get the current font size
  fpToFloat(kryonCanvasGetFontSize())

# ============================================================================
# Convenience Functions (Love2D style)
# ============================================================================

proc background*(color: uint32) =
  ## Set background color and clear canvas
  setBackgroundColor(color)
  clearCanvas()

proc background*(r, g, b, a: int) =
  ## Set background color and clear canvas (RGBA)
  setBackgroundColor(r, g, b, a)
  clearCanvas()

proc background*(r, g, b: int) =
  ## Set background color and clear canvas (RGB)
  setBackgroundColor(r, g, b, 255)
  clearCanvas()

proc fill*(color: uint32) =
  ## Set fill color
  setColor(color)

proc fill*(r, g, b, a: int) =
  ## Set fill color (RGBA)
  setColor(r, g, b, a)

proc fill*(r, g, b: int) =
  ## Set fill color (RGB)
  setColor(r, g, b, 255)

proc stroke*(color: uint32) =
  ## Set stroke color
  setColor(color)

proc stroke*(r, g, b, a: int) =
  ## Set stroke color (RGBA)
  setColor(r, g, b, a)

proc stroke*(r, g, b: int) =
  ## Set stroke color (RGB)
  setColor(r, g, b, 255)

proc strokeWeight*(weight: float) =
  ## Set stroke weight (line width)
  setLineWidth(weight)

proc noFill*() =
  ## Disable fill (draw outlines only)
  setColor(0) # Transparent color

proc noStroke*() =
  ## Disable stroke (no outlines)
  setLineWidth(0)

# ============================================================================
# Shape Drawing Convenience Functions
# ============================================================================

proc rect*(x, y, width, height: float) =
  ## Draw a filled rectangle
  rectangle(DrawMode.Fill, x, y, width, height)

proc rect*(x, y, width, height: int) =
  ## Draw a filled rectangle - integer version
  rectangle(DrawMode.Fill, x, y, width, height)

proc rectMode*() =
  ## Rectangle drawing mode (placeholder for compatibility)
  discard

proc ellipseMode*() =
  ## Ellipse drawing mode (placeholder for compatibility)
  discard

# ============================================================================
# Color Utilities
# ============================================================================

proc color*(r, g, b, a: int): uint32 =
  ## Create color from RGBA values
  rgb(r, g, b, a)

proc color*(r, g, b: int): uint32 =
  ## Create color from RGB values
  rgb(r, g, b, 255)

proc color*(gray: int): uint32 =
  ## Create grayscale color
  rgb(gray, gray, gray, 255)

proc color*(gray, alpha: int): uint32 =
  ## Create grayscale color with alpha
  rgb(gray, gray, gray, alpha)

proc red*(color: uint32): int =
  ## Get red component from color
  int(colorRed(color))

proc green*(color: uint32): int =
  ## Get green component from color
  int(colorGreen(color))

proc blue*(color: uint32): int =
  ## Get blue component from color
  int(colorBlue(color))

proc alpha*(color: uint32): int =
  ## Get alpha component from color
  int(colorAlpha(color))

# ============================================================================
# Math Utilities
# ============================================================================

proc radians*(degrees: float): float =
  ## Convert degrees to radians
  degrees * PI / 180.0

proc degrees*(radians: float): float =
  ## Convert radians to degrees
  radians * 180.0 / PI

proc random*(low, high: float): float =
  ## Generate random float between low and high
  # Simple implementation - in real use would use proper random generator
  low + (high - low) * 0.5

proc random*(max: float): float =
  ## Generate random float between 0 and max
  random(0.0, max)

proc map*(value, inMin, inMax, outMin, outMax: float): float =
  ## Map value from one range to another
  (value - inMin) / (inMax - inMin) * (outMax - outMin) + outMin

# ============================================================================
# Advanced Drawing Functions
# ============================================================================

proc drawHeart*(x, y, size: float) =
  ## Draw a heart shape at the specified position
  push()
  translate(x, y)
  scale(size / 50.0, size / 50.0)

  # Heart shape using bezier curves approximation
  setColor(255, 0, 0, 255)

  # Draw using circles and transforms for simple heart
  push()
  translate(-20, -15)
  circle(Fill, 0, 0, 15)
  translate(40, 0)
  circle(Fill, 0, 0, 15)
  pop()

  # Bottom triangle
  polygon(Fill, [
    (0.0, 10.0),    # Bottom point
    (-25.0, -5.0),  # Left top
    (25.0, -5.0)    # Right top
  ])

  pop()

proc drawStar*(x, y, size: float, points: int = 5) =
  ## Draw a star shape
  if points < 3: return

  push()
  translate(x, y)
  scale(size / 50.0, size / 50.0)

  var vertices = newSeq[Point](points * 2)

  for i in 0..<points:
    let angle1 = float(i * 360 / points)
    let angle2 = float((i * 2 + 1) * 180 / points)

    # Outer point
    vertices[i * 2] = (
      cos(radians(angle1)) * 25.0,
      sin(radians(angle1)) * 25.0
    )

    # Inner point
    vertices[i * 2 + 1] = (
      cos(radians(angle2)) * 10.0,
      sin(radians(angle2)) * 10.0
    )

  polygon(Fill, vertices)
  pop()

# ============================================================================
# Template for Common Patterns
# ============================================================================

template withTransform*(body: untyped) =
  ## Execute drawing code with pushed transform stack
  push()
  body
  pop()

template withColor*(color: uint32, body: untyped) =
  ## Execute drawing code with a specific color, then restore original
  let oldColor = getColor()
  setColor(color)
  body
  setColor(oldColor)

template withStyle*(strokeColor: uint32 = 0, fillColor: uint32 = 0,
                   lineWidth: float = 1.0, body: untyped) =
  ## Execute drawing code with specific style, then restore
  let oldColor = getColor()
  let oldLineWidth = getLineWidth()

  if strokeColor != 0:
    setColor(strokeColor)
  if fillColor != 0:
    setColor(fillColor)
  if lineWidth > 0:
    setLineWidth(lineWidth)

  body
  setColor(oldColor)
  setLineWidth(oldLineWidth)

File: bindings/lua/canvas.lua
-- Kryon Canvas API - Love2D Style for Lua
--
-- Provides Love2D-inspired immediate mode drawing functions
-- that map to the enhanced C canvas implementation
--
-- Usage:
--   local canvas = require("canvas")
--   canvas.background(20, 20, 30)
--   canvas.fill(255, 100, 100)
--   canvas.rectangle("fill", 50, 50, 100, 80)

local M = {}

-- Import low-level canvas bindings
local kryon = require("kryon")

-- ============================================================================
-- Canvas State Management
-- ============================================================================

-- Current drawing state
local state = {
    fill_color = {255, 255, 255, 255},
    stroke_color = {255, 255, 255, 255},
    stroke_weight = 1,
    line_style = "solid",
    blend_mode = "alpha",
    font_size = 16,
    current_transform = {
        x = 0, y = 0,
        rotation = 0,
        scale_x = 1, scale_y = 1
    }
}

-- Transform stack for push/pop operations
local transform_stack = {}

-- ============================================================================
-- Color Management Functions
-- ============================================================================

function M.color(r, g, b, a)
    a = a or 255
    return kryon.color_rgba(r, g, b, a)
end

function M.fill(r, g, b, a)
    if type(r) == "table" then
        -- Support M.fill({r, g, b, a}) format
        state.fill_color = {r[1] or 255, r[2] or 255, r[3] or 255, r[4] or 255}
    else
        state.fill_color = {r or 255, g or 255, b or 255, a or 255}
    end
end

function M.stroke(r, g, b, a)
    if type(r) == "table" then
        -- Support M.stroke({r, g, b, a}) format
        state.stroke_color = {r[1] or 255, r[2] or 255, r[3] or 255, r[4] or 255}
    else
        state.stroke_color = {r or 255, g or 255, b or 255, a or 255}
    end
end

function M.strokeWeight(width)
    state.stroke_weight = width or 1
end

function M.background(r, g, b, a)
    -- Clear the entire canvas with background color
    local bg_color
    if type(r) == "table" then
        bg_color = M.color(r[1], r[2], r[3], r[4])
    else
        bg_color = M.color(r or 0, g or 0, b or 0, a or 255)
    end

    -- This would call the C canvas clear function
    -- TODO: Implement canvas clear in C bindings
    print("Setting background to:", string.format("0x%08X", bg_color))
end

function M.clear()
    -- Clear canvas to transparent
    M.background(0, 0, 0, 0)
end

-- ============================================================================
-- Color Constants (matching Nim version)
-- ============================================================================

M.COLOR_RED = M.color(255, 0, 0, 255)
M.COLOR_GREEN = M.color(0, 255, 0, 255)
M.COLOR_BLUE = M.color(0, 0, 255, 255)
M.COLOR_YELLOW = M.color(255, 255, 0, 255)
M.COLOR_CYAN = M.color(0, 255, 255, 255)
M.COLOR_MAGENTA = M.color(255, 0, 255, 255)
M.COLOR_ORANGE = M.color(255, 165, 0, 255)
M.COLOR_PURPLE = M.color(128, 0, 128, 255)
M.COLOR_WHITE = M.color(255, 255, 255, 255)
M.COLOR_BLACK = M.color(0, 0, 0, 255)
M.COLOR_TRANSPARENT = M.color(0, 0, 0, 0)

-- ============================================================================
-- Drawing Mode Constants
-- ============================================================================

M.FILL = "fill"
M.LINE = "line"
M.FILL_AND_LINE = "fill_and_line"

-- ============================================================================
-- Shape Drawing Functions
-- ============================================================================

function M.rectangle(mode, x, y, width, height)
    local color = mode == "fill" or mode == "fill_and_line" and state.fill_color or state.stroke_color
    local final_color = M.color(color[1], color[2], color[3], color[4])

    -- TODO: Call C canvas rectangle function
    print(string.format("Drawing rectangle: mode=%s, pos=(%d,%d), size=(%d,%d), color=0x%08X",
          mode, x, y, width, height, final_color))
end

function M.circle(mode, x, y, radius)
    local color = mode == "fill" or mode == "fill_and_line" and state.fill_color or state.stroke_color
    local final_color = M.color(color[1], color[2], color[3], color[4])

    -- TODO: Call C canvas circle function
    print(string.format("Drawing circle: mode=%s, center=(%d,%d), radius=%d, color=0x%08X",
          mode, x, y, radius, final_color))
end

function M.ellipse(mode, x, y, radius_x, radius_y)
    local color = mode == "fill" or mode == "fill_and_line" and state.fill_color or state.stroke_color
    local final_color = M.color(color[1], color[2], color[3], color[4])

    -- TODO: Call C canvas ellipse function
    print(string.format("Drawing ellipse: mode=%s, center=(%d,%d), radii=(%d,%d), color=0x%08X",
          mode, x, y, radius_x, radius_y, final_color))
end

function M.line(x1, y1, x2, y2)
    local color = state.stroke_color
    local final_color = M.color(color[1], color[2], color[3], color[4])

    -- TODO: Call C canvas line function
    print(string.format("Drawing line: from=(%d,%d) to=(%d,%d), width=%d, color=0x%08X",
          x1, y1, x2, y2, state.stroke_weight, final_color))
end

function M.line(points)
    if type(points) == "table" and #points >= 2 then
        for i = 1, #points - 1 do
            local p1 = points[i]
            local p2 = points[i + 1]
            M.line(p1.x or p1[1], p1.y or p1[2], p2.x or p2[1], p2.y or p2[2])
        end
    end
end

function M.point(x, y)
    local color = state.fill_color
    local final_color = M.color(color[1], color[2], color[3], color[4])

    -- TODO: Call C canvas point function
    print(string.format("Drawing point: (%d,%d), color=0x%08X", x, y, final_color))
end

function M.polygon(mode, points)
    local color = mode == "fill" or mode == "fill_and_line" and state.fill_color or state.stroke_color
    local final_color = M.color(color[1], color[2], color[3], color[4])

    -- Convert points to expected format
    local point_list = {}
    for _, point in ipairs(points) do
        table.insert(point_list, {x = point.x or point[1], y = point.y or point[2]})
    end

    -- TODO: Call C canvas polygon function
    print(string.format("Drawing polygon: mode=%s, points=%d, color=0x%08X",
          mode, #point_list, final_color))
end

function M.arc(mode, x, y, radius, start_angle, end_angle)
    local color = mode == "fill" or mode == "fill_and_line" and state.fill_color or state.stroke_color
    local final_color = M.color(color[1], color[2], color[3], color[4])

    -- TODO: Call C canvas arc function
    print(string.format("Drawing arc: mode=%s, center=(%d,%d), radius=%d, angles=(%.2f,%.2f), color=0x%08X",
          mode, x, y, radius, start_angle, end_angle, final_color))
end

-- ============================================================================
-- Transform Functions
-- ============================================================================

function M.push()
    -- Save current transform state
    table.insert(transform_stack, {
        x = state.current_transform.x,
        y = state.current_transform.y,
        rotation = state.current_transform.rotation,
        scale_x = state.current_transform.scale_x,
        scale_y = state.current_transform.scale_y
    })
end

function M.pop()
    -- Restore previous transform state
    if #transform_stack > 0 then
        local saved_transform = table.remove(transform_stack)
        state.current_transform = saved_transform

        -- TODO: Apply transform to canvas in C bindings
        print("Restored transform")
    end
end

function M.translate(x, y)
    state.current_transform.x = state.current_transform.x + x
    state.current_transform.y = state.current_transform.y + y

    -- TODO: Apply translation to canvas in C bindings
    print(string.format("Translate: (%d, %d)", x, y))
end

function M.rotate(angle)
    state.current_transform.rotation = state.current_transform.rotation + angle

    -- TODO: Apply rotation to canvas in C bindings
    print(string.format("Rotate: %.2f degrees", angle))
end

function M.scale(x, y)
    y = y or x  -- Default to uniform scaling
    state.current_transform.scale_x = state.current_transform.scale_x * x
    state.current_transform.scale_y = state.current_transform.scale_y * y

    -- TODO: Apply scale to canvas in C bindings
    print(string.format("Scale: (%.2f, %.2f)", x, y))
end

-- ============================================================================
-- Text Drawing Functions
-- ============================================================================

function M.print(text, x, y)
    local color = state.fill_color
    local final_color = M.color(color[1], color[2], color[3], color[4])

    -- TODO: Call C canvas text function
    print(string.format("Drawing text: '%s' at (%d,%d), size=%d, color=0x%08X",
          text, x, y, state.font_size, final_color))
end

function M.printf(text, x, y, limit, align)
    align = align or "left"

    -- TODO: Implement formatted text with alignment in C bindings
    print(string.format("Drawing formatted text: '%s' at (%d,%d), limit=%d, align=%s",
          text, x, y, limit, align))
end

function M.setFont(font)
    -- TODO: Set font in C bindings
    print("Setting font:", font)
end

function M.setFontSize(size)
    state.font_size = size
    print("Setting font size:", size)
end

-- ============================================================================
-- Convenience Functions
-- ============================================================================

function M.withColor(color, func)
    local old_fill = {table.unpack(state.fill_color)}
    local old_stroke = {table.unpack(state.stroke_color)}

    M.fill(color)
    M.stroke(color)

    local result = func()

    state.fill_color = old_fill
    state.stroke_color = old_stroke

    return result
end

function M.withTransform(func)
    M.push()
    local result = func()
    M.pop()
    return result
end

-- ============================================================================
-- Special Shape Functions
-- ============================================================================

function M.drawHeart(x, y, size)
    M.withTransform(function()
        M.translate(x, y)
        M.scale(size / 40, size / 40)

        -- Draw heart shape using two circles and a polygon
        M.fill(255, 100, 150)  -- Pink
        M.circle("fill", -15, -15, 15)  -- Left circle
        M.circle("fill", 15, -15, 15)   -- Right circle

        -- Bottom triangle
        local heart_points = {
            {x = -30, y = 0},
            {x = 30, y = 0},
            {x = 0, y = 35}
        }
        M.polygon("fill", heart_points)
    end)
end

function M.drawStar(x, y, size, points)
    points = points or 5

    M.withTransform(function()
        M.translate(x, y)
        M.scale(size / 30, size / 30)

        local star_points = {}
        for i = 0, points * 2 - 1 do
            local angle = (i * math.pi) / points
            local radius
            if i % 2 == 0 then
                radius = 30  -- Outer radius
            else
                radius = 12  -- Inner radius
            end

            table.insert(star_points, {
                x = math.cos(angle - math.pi / 2) * radius,
                y = math.sin(angle - math.pi / 2) * radius
            })
        end

        M.fill(255, 255, 100)  -- Yellow
        M.polygon("fill", star_points)
    end)
end

-- ============================================================================
-- Canvas State Reset
-- ============================================================================

function M.reset()
    state.fill_color = {255, 255, 255, 255}
    state.stroke_color = {255, 255, 255, 255}
    state.stroke_weight = 1
    state.line_style = "solid"
    state.blend_mode = "alpha"
    state.font_size = 16
    state.current_transform = {
        x = 0, y = 0,
        rotation = 0,
        scale_x = 1, scale_y = 1
    }
    transform_stack = {}
end

return M